redis/src/main/java/org/hibernate/ogm/datastore/redis/RedisJsonDialect.java

```diff
@@ -361,16 +365,64 @@ public void forEachTuple(final ModelConsumer consumer, TupleTypeContext tupleTyp
 		ScanArgs scanArgs = ScanArgs.Builder.matches( prefix + "*" );
 		do {
 			cursor = scan( cursor, scanArgs );
+			consumer.consume( new RedisJsonDialectTupleSupplier( cursor, entityStorageStrategy, prefix, entityKeyMetadata ) );
+		} while ( !cursor.isFinished() );
+	}
 
-			for ( String key : cursor.getKeys() ) {
-				Entity document = entityStorageStrategy.getEntity( key );
+	private class RedisJsonDialectTupleSupplier implements TupleSupplier {
 
-				addKeyValuesFromKeyName( entityKeyMetadata, prefix, key, document );
+		private final KeyScanCursor<String> cursor;
+		private final String prefix;
+		private final EntityKeyMetadata entityKeyMetadata;
+		private final JsonEntityStorageStrategy storageStrategy;
 
-				consumer.consume( new Tuple( new RedisJsonTupleSnapshot( document ), SnapshotType.UPDATE ) );
-			}
+		public RedisJsonDialectTupleSupplier(KeyScanCursor<String> cursor, JsonEntityStorageStrategy storageStrategy, String prefix, EntityKeyMetadata entityKeyMetadata) {
+			this.cursor = cursor;
+			this.storageStrategy = storageStrategy;
+			this.prefix = prefix;
+			this.entityKeyMetadata = entityKeyMetadata;
+		}
 
-		} while ( !cursor.isFinished() );
+		@Override
+		public ClosableIterator<Tuple> get(TransactionContext transactionContext) {
+			return new RedisJsonTupleIterator( cursor, storageStrategy, prefix, entityKeyMetadata );
+		}
+	}
+
+	private class RedisJsonTupleIterator implements ClosableIterator<Tuple> {
+
+		private final Iterator<String> iterator;
+		private final EntityKeyMetadata entityKeyMetadata;
+		private final String prefix;
+		private final JsonEntityStorageStrategy storageStrategy;
+
+		public RedisJsonTupleIterator(KeyScanCursor<String> cursor, JsonEntityStorageStrategy storageStrategy, String prefix, EntityKeyMetadata entityKeyMetadata) {
+			this.storageStrategy = storageStrategy;
+			this.prefix = prefix;
+			this.entityKeyMetadata = entityKeyMetadata;
+			this.iterator = cursor.getKeys().iterator();
+		}
+
+		@Override
+		public boolean hasNext() {
+			return iterator.hasNext();
+		}
+
+		@Override
+		public Tuple next() {
+			String key = iterator.next();
```

Comment:
I think this gives us the ability to do some batching on entity loading. WDYT about it?

Should be pursued as another issue, of course.